WordPress/gutenberg
# Deprecate interface package's preference APIs

Part of https://github.com/WordPress/gutenberg/issues/31965

## What?
Deprecates the feature preference actions and selectors in the interface store and makes them select and dispatch from the new preferences package.

## Why?
In #38873 a new package was introduced which replaces what these selectors and actions do.

The editors that were previously using the interface package for feature preferences have already been migrated to use the new preferences package (#39084, #39112, #39115).

These old APIs now need to be deprecated, but they need to also still work for any third party usage.

## How?
- Removes the interface package's preference reducer and persistence of that reducer
- Adds deprecation messages to the selectors and actions
- Updated actions to dispatch to the preference package instead of the interface package
- Updates the one selector to select from the preferences package instead of the interface package
- Migrates the local storage data for any non-core feature preferences, so that there's no break in functionality if a plugin is using this (which is unlikely).

## Testing Instructions
There's no user facing way to test this, the easiest option is to use the API in the browser console the way a plugin would.

1. Checkout `trunk` and build
2. Open up the post editor
3. In the browser console run the following commands:
```js
// Set some defaults
wp.data.dispatch( 'core/interface' ).setFeatureDefaults( 'my-test-plugin', { a: true, b: false } );
// Toggle 'b' to true
wp.data.dispatch( 'core/interface' ).toggleFeature( 'my-test-plugin', 'b' );
```
4. Checkout this branch and build
5. In the browser console run the following commands:
```js
// Set the defaults again.
wp.data.dispatch( 'core/interface' ).setFeatureDefaults( 'my-test-plugin', { a: true, b: false } );
// 'a' should be true due to the default value
wp.data.select( 'core/interface' ).isFeatureActive( 'my-test-plugin', 'a' ); // should be `true`.
// 'b' should be true as it was set to true in step three.
wp.data.select( 'core/interface' ).isFeatureActive( 'my-test-plugin', 'b' ); // should be `true`.
// This can be accessed directly via the preferences store, too.
wp.data.select( 'core/preferences' ).get( 'my-test-plugin', 'b' ); // should be `true`.
```
